tests/store/artifact/test_cli.py

```diff
@@ -71,21 +82,61 @@ def test_get_artifact_repository(artifact_uri):
-                f.write("test")
-            mlflow.log_artifact(local_path, "test")
-    command = ["mlflow", "artifacts", "download", "-u"]
-    # Test with run uri
-    run_uri = "runs:/{run_id}/test".format(run_id=run.info.run_id)
-    actual_uri = posixpath.join(run.info.artifact_uri, "test")
+def _run_download_artifact_command(command):
+    """
+    :param command: An `mlflow artifacts` command list.
+    :return: The downloaded artifact content.
+    """
+    p = Popen(command, stdout=PIPE, stderr=STDOUT)
+    output = p.stdout.readlines()
+    download_output_path = pathlib.Path(output[-1].strip().decode("utf-8"))
+    downloaded_file = list(download_output_path.iterdir())[0]
+    return downloaded_file.read_text()
+
+
+def test_download_artifacts_with_uri(run_with_artifact):
+    run, artifact_path, artifact_content = run_with_artifact
+    base_command = ["mlflow", "artifacts", "download", "-u"]
+    run_uri = f"runs:/{run.info.run_id}/{artifact_path}"
+    actual_uri = pathlib.PurePosixPath(run.info.artifact_uri) / artifact_path
     for uri in (run_uri, actual_uri):
-        p = Popen(command + [uri], stdout=PIPE, stderr=STDOUT)
-        output = p.stdout.readlines()
-        downloaded_file_path = output[-1].strip()
-        downloaded_file = os.listdir(downloaded_file_path)[0]
-        with open(os.path.join(downloaded_file_path, downloaded_file), "r") as f:
-            assert f.read() == "test"
+        downloaded_content = _run_download_artifact_command(base_command + [uri])
+        assert downloaded_content == artifact_content
+
+    # Check for backwards compatibility with preexisting behavior in MLflow <= 1.24.0 where
+    # specifying `artifact_uri` and `artifact_path` together did not throw an exception (unlike
+    # `mlflow.artifacts.download_artifacts()`) and instead used `artifact_uri` while ignoring
+    # `run_id` and `artifact_path`
+    downloaded_content = _run_download_artifact_command(
+        base_command + [uri] + ["--run-id", "bad", "--artifact-path", "bad"]
+    )
+    assert downloaded_content == artifact_content
+
+
+def test_download_artifacts_with_run_id_and_path(run_with_artifact):
+    run, artifact_path, artifact_content = run_with_artifact
+    downloaded_content = _run_download_artifact_command(
+        [
+            "mlflow",
+            "artifacts",
+            "download",
+            "--run-id",
+            run.info.run_id,
+            "--artifact-path",
+            artifact_path,
+        ]
+    )
+    assert downloaded_content == artifact_content
+
+
+@pytest.mark.parametrize("dst_subdir_path", [None, "doesnt_exist_yet"])
+def test_download_artifacts_with_dst_path(run_with_artifact, tmp_path, dst_subdir_path):
+    run, artifact_path, _ = run_with_artifact
+    artifact_uri = f"runs:/{run.info.run_id}/{artifact_path}"
+    dst_path = tmp_path / dst_subdir_path if dst_subdir_path else tmp_path
+
+    command = ["mlflow", "artifacts", "download", "-u", artifact_uri, "-d", dst_path]
```

Comment:
It appears Popen doesn't accept a command args containing a path-like object in Python 3.7:

https://github.com/mlflow/mlflow/runs/5796393850?check_suite_focus=true#step:6:2081

```
>           needquote = (" " in arg) or ("\t" in arg) or not arg
E           TypeError: argument of type 'WindowsPath' is not iterable
```

https://docs.python.org/3/library/subprocess.html#subprocess.Popen says:

> Changed in version 3.8: args parameter accepts a [path-like object](https://docs.python.org/3/glossary.html#term-path-like-object) if shell is False and a sequence containing bytes and path-like objects on Windows.

Reply:
will it work with a `str(dst_path)` ?

Reply:
Yes, that should work

Reply:
Thanks folks! Fixed!